### DISCUSSION TITLE
Helmfile support for helm rollback

### DISCUSSION THREAD

**Discussion by MichaelMorrisEst** [helmfile/helmfile#1064] [Ideas]
Hi,

I am interesting in contributing a rollback feature to helmfile and would like to get the thoughts of the community before I start implementing anything to see if people think this would be a good idea, how it would work and any potential problems.
As a starting point I would like to propose:

- User calls 'helmfile apply'
- For each release where the chart version is less than the chart version already on the cluster then call helm rollback for that release
- For each release where this is not the case, then proceed as today

This would enable rollback specific hooks to be invoked when a release is being rolled back to a previous version.

Any thoughts/comments would be appreciated.

**Comment by yxxhero** [comment]
one question:
 why do we need rollback? are there any user cases?

**Reply by MichaelMorrisEst** [reply]
Use case is as follows:
- 'helmfile apply' is invoked
-  errors result for some releases
- user wants to go back to previous chart versions to resolve the errors
- user steps back the chart versions and invokes 'helmfile apply' again
- However
    - there may be actions that are required to be performed when stepping back the chart versions
    - these are typically included in helm rollback hooks
    - there is currently no way to step back the chart version using helmfile that will result the rollback hooks being invoked, and therefore helmfile cannot be used to perform this activity

**Reply by yxxhero** [reply]
@MichaelMorrisEst That sounds like a violation of declarative logic. I think you should rollback the helmfile configuration when applying errors.  then apply again. 

/cc @mumoshu

**Reply by MichaelMorrisEst** [reply]
@yxxhero thanks for your input. Are you saying that using 'helm rollback' is a violation of declarative logic? There is currently no way to execute a 'helm rollback' through helmfile. We are just looking for a way to do this as we have scenarios that require rollback hooks when stepping back image versions. Using 'helm apply' will work in most cases but where it needs to be a 'helm rollback' then we are stuck

**Reply by xiaomudk** [reply]
@MichaelMorrisEst First of all, thank you very much for your question.
Can you give a specific helmfile example?
I agree with yxxhero’s opinion. If `helmfile apply` cannot be directly updated to the old version, is there something wrong with helmfile.yaml?

**Comment by MichaelMorrisEst** [comment]
I have created a PR to show in more detail what I have in mind. I have added a flag to switch between the new behaviour and existing behaviour to ensure backwards compatibility.
The logic when the new flag is set is proposed to be as follows:

- A 'helm rollback' to be performed when the following conditions are met:
   - chart version is provided in the helmfile
   - chart version specified in the helmfile is less than the chart version of the latest release revision
   - a previous release revision is found with the same chart version as in the helmfile
- The rollback shall be to the highest release revision with the same chart version as is given in the helmfile
- A 'helm upgrade' shall be performed (as today) where any of the above conditions are not met

Any thoughts/comments/concerns appreciated
